pipelines/pipelines.go

```diff
@@ -0,0 +1,198 @@
+func FlatMapCtx[S, T any](ctx context.Context, in <-chan S, f func(context.Context, S) []T) <-chan T {
+    tChan := make(chan T)
+
+    go func() {
+        for ss := range in {
+            select {
+            case <-ctx.Done():
+                return
+            default:
+                SendAll(ctx, f(ctx, ss), tChan)
+            }
+
+        }
+        close(tChan)
+    }()
+    return tChan
+}
+
+// Combine combines the values from two channels into a third, which is returned. The returned channel is closed once
+// either of the input channels are closed, or the provided context is cancelled.
+func Combine[T any](ctx context.Context, s1 <-chan T, s2 <-chan T) chan T {
+    out := make(chan T)
+    var wg sync.WaitGroup
+    wg.Add(2)
+    go func() {
+        defer wg.Done()
+        for u := range s1 {
+            select {
+            case <-ctx.Done():
+                return
+            case out <- u:
+            }
+
+        }
+    }()
+    go func() {
+        defer wg.Done()
+        for u := range s2 {
+            select {
+            case <-ctx.Done():
+                return
+            case out <- u:
+            }
+        }
+    }()
+    go func() {
+        wg.Wait()
+        close(out)
+    }()
+    return out
+}
+
+// WithCancel returns a new channel which receives each value from ch. If the provided context is cancelled or the
+// input channel is closed, the returned channel is also closed.
+func WithCancel[T any](ctx context.Context, ch <-chan T) <-chan T {
+    result := make(chan T)
+    go func() {
+        for u := range ch {
+            select {
+            case <-ctx.Done():
+                return
+            case result <- u:
+            }
+        }
+        close(result)
+    }()
+    return result
+}
+
+// SendAll sends all values in ts to the provided channel while handling context closure. It blocks until the channel
+// is closed, or the provided context is cancelled.
+func SendAll[T any](ctx context.Context, ts []T, ch chan<- T) {
+    for _, t := range ts {
+        select {
+        case <-ctx.Done():
+            return
+        case ch <- t:
+        }
+    }
+}
+
+// ForkMapCtx starts a pipeline stage which, for each value of T received from in, forks an invocation of f onto a
+// new goroutine. Any values sent to the channel provided to f are serialized and made available in the output channel.
+func ForkMapCtx[S, T any](ctx context.Context, in <-chan S, f func(context.Context, S, chan<- T)) <-chan T {
+    tChan := make(chan T)
+    var wg sync.WaitGroup
+    go func() {
+        for s := range in {
+            select {
+            case <-ctx.Done():
+                return
+            default:
+                wg.Add(1)
+                go func(s S) {
+                    f(ctx, s, tChan)
+                    wg.Done()
+                }(s)
+            }
+        }
+        wg.Wait()
+        close(tChan)
+    }()
+    return tChan
+}
```

Comment:
Need a contextless version of ForkMap...

Reply:
On second thought, no we don't. Context should always be included here to encourage authors to respect context cancellation when sending values to the output channel -- otherwise they could leak a goroutine.